lib/services/local/dynamic_access.go

```diff
@@ -69,23 +72,38 @@ func (s *DynamicAccessService) SetAccessRequestState(ctx context.Context, name s
 	// to be "first come first serve".  Denials should overwrite approvals, but
 	// approvals should not overwrite denials.
 	for i := 0; i < maxCmpAttempts; i++ {
-		item, err := s.Get(ctx, accessRequestKey(name))
+		item, err := s.Get(ctx, accessRequestKey(params.RequestID))
 		if err != nil {
 			if trace.IsNotFound(err) {
-				return trace.NotFound("cannot set state of access request %q (not found)", name)
+				return trace.NotFound("cannot set state of access request %q (not found)", params.RequestID)
 			}
 			return trace.Wrap(err)
 		}
 		req, err := itemToAccessRequest(*item)
 		if err != nil {
 			return trace.Wrap(err)
 		}
-		if err := req.SetState(state); err != nil {
+		if err := req.SetState(params.State); err != nil {
 			return trace.Wrap(err)
 		}
+		req.SetResolveReason(params.Reason)
+		req.SetResolveAnnotations(params.Annotations)
+		if len(params.Roles) > 0 {
+		Outer:
```

Comment:
Can you pull this out into it's own function. I think loops and labels like this are confusing.